Creating-Objects.md

```diff
@@ -1,12 +1,105 @@
+## Table of Contents
+
+* [Introduction](#introduction)
+* [How objects are defined](#how-objects-are-defined)
+  * [Python classes](#python-classes)
+  * [Angular components](#angular-components)
+* [Create a new object](#create-a-new-object)
+* [Using objects](#using-objects)
+
+## Introduction
+
+When collecting user input, we often define the kind of input we expect using [[schemas|Schemas]]. These schemas can describe arbitrarily nested collections of base types we call _objects_. For example, consider this schema:
+
+```json
+{
+  "type": "dict",
+  "properties": [
+    {
+      "name": "version",
+      "schema": {
+        "type": "int"
+      }
+    }
+  ]
+}
+```
+
+The schema would match a dictionary like this:
+
+```json
+{"version": 34}
+```
+
+The `dict` type is a collection, but the `int` type is an object. We define these objects as classes in [`extensions/objects/models/objects.py`](https://github.com/oppia/oppia/blob/develop/extensions/objects/models/objects.py) and components in [`extensions.objects/templates`](https://github.com/oppia/oppia/blob/develop/extensions/objects/templates).
+
+## How objects are defined
+
+### Python classes
+
+The Python classes for objects are defined in `objects.py`. For example, here's the class for the `int` type:
+
+```python
+class Int(BaseObject):
+    """Integer class."""
+
+    description = 'An integer.'
+    default_value = 0
+
+    @classmethod
+    def get_schema(cls):
+        """Returns the object schema.
+
+        Returns:
+            dict. The object schema.
+        """
+        return {
+            'type': 'int'
+        }
+```
+
+Object classes all meet the following criteria:
+
+* Inherit from either `BaseObject` or another object class.
+* Set the class variable `description` to a string describing the object.
+* Set the class variable `default_value` to the value that should be the default whenever this type is used in a rule. Note that this variable is technically not required if the object will never be used in a [[rule|Creating-rules]].
+* Provide a `normalize` class method that accepts a raw Python object. The method should first validate the raw object to check whether it is malformed. If the validation checks pass, then the method should return a normalized form of the object. Note that in this context, "normalized" means a primitive Python type or multiple primitive types combined with lists and/or dicts.
+
+  * The `BaseObject` class provides a `normalize` class method that can handle normalization for you so long as your object class provides a `get_schema` class method that returns a schema describing the object. `BaseObject.normalize` passes your schema to `schema_utils.normalize_against_schema`, so make sure that `normalize_against_schema` won't try to normalize your object to your schema using your object class. Otherwise, you'll get an endless loop between `normalize_against_schema` and `BaseObject.normalize`.
```

Comment:
Probably not "your object"? The "object class" seems reasonable as "your" since the reader is presumably writing that, but the objects that come into the normalize() method are external ones over which the reader has no control.

Reply:
done